### PR TITLE
Ranking UI foundation

### PR DESCRIPTION
**Pull request by CHOOSEIT**
This PR aims to lays the foundations for the tasks #298 and #299.

It provides and test two data classes `RankingElementDetails` and `RankingDetails` that contains the necessary information needed to display the ranking page.

`RankingElementDetails` provide the necessary information to an entry in the ranking page:
image

`RankingDetails` contains a list of `RankingElementDetails` along with the `RankingElementDetails` information about the current user.
image

Theses images comes from our mockup

### CODE REVIEW

File: lib/models/ui/ranking/ranking_details.dart [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+/// to the developer. Moreover, when used in the UI, this will yield a runtime error
+/// if the [userRank] is null.
+@immutable
+class RankingDetails {
+  const RankingDetails({
+    required this.userRankElementDetails,
+    required this.rankElementDetailsList,
+  });
+
+  /// Details of the user's rank.
```
Comment (gruvw) [inline lib/models/ui/ranking/ranking_details.dart:17]: Can you add here (in the documentation) the fact that the `userRank` attribute of this `RankingElementDetails` will be null.

File: lib/models/ui/ranking/ranking_details.dart [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  final RankingElementDetails userRankElementDetails;
+
+  /// List containing details the ranking elements.
+  final List<RankingElementDetails> rankElementDetailsList;
```
Comment (gruvw) [inline lib/models/ui/ranking/ranking_details.dart:21]: Can you add here (in the documentation) the fact that the `userRank` attribute of every `RankingElementDetails` in this list wont be null.

File: lib/models/ui/ranking/ranking_element_details.dart [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+/// The [userRank] is nullable to allow the current user to not have a rank.
+@immutable
+class RankingElementDetails {
+  const RankingElementDetails({
+    required this.userDisplayName,
+    required this.userUserName,
+    required this.centauriPoints,
+    required this.userRank,
+  });
+
```
Comment (gruvw) [inline lib/models/ui/ranking/ranking_element_details.dart:13]: I think here we should have the user id too. We might use it for the avatar and the profile pop up.

**Review by gruvw — CHANGES_REQUESTED** [review]

**Comment by gruvw** [comment]
Great start on this PR, a just noticed a few things to change.

File: lib/models/ui/ranking/ranking_details.dart (hunk 1 above)
Comment (CHOOSEIT) [reply lib/models/ui/ranking/ranking_details.dart:17]: You are completely right !

Added in: [8a55765](https://github.com/ProximaEPFL/proxima/pull/319/commits/8a55765d5958a96c8109fae1ca8da60b472f64cf)

Thank you !

File: lib/models/ui/ranking/ranking_details.dart (hunk 2 above)
Comment (CHOOSEIT) [reply lib/models/ui/ranking/ranking_details.dart:24]: You are completely right !

Added in: [8a55765](https://github.com/ProximaEPFL/proxima/pull/319/commits/8a55765d5958a96c8109fae1ca8da60b472f64cf)

**Note:** that I also added assertion test on the constructor to ensure that the `userRank` are non-null, because I realised that in the UI, I wont test it and it may not yield an error.

Thank you !

File: lib/models/ui/ranking/ranking_element_details.dart (hunk 3 above)
Comment (CHOOSEIT) [reply lib/models/ui/ranking/ranking_element_details.dart:13]: I've made some local progress on the user interface and it seems that the for the moment the `userId` is not necessary and I am able to display the ranking card information along with the pop up.

I will follow up on this discussion, if I have any update :)

Thank you :)

File: lib/models/ui/ranking/ranking_details.dart (hunk 2 above)
Comment (gruvw) [reply lib/models/ui/ranking/ranking_details.dart:24]: Asserts are indeed a great addition, thank you :)

File: lib/models/ui/ranking/ranking_element_details.dart (hunk 3 above)
Comment (gruvw) [reply lib/models/ui/ranking/ranking_element_details.dart:13]: Oh right. I just double-checked, I thought that `UserProfilePopUp` was taking a user id as parameter but apparently not :+1:

**Review by gruvw — APPROVED** [review]
Awesome start :) We can now both work on our task #298 and #299 on our own branch from this common base.

We'll merge this when our 2 other PRs are ready to merge.
